// smartface/lib/log/log4j.js
/**
 * Logger for creating same level logger.
 * @file log4j.js
 * @version 1.0.0
 * @requires module:fs
 * @requires module:underscore
 * @requires module:log4js
 */
const fs = require('fs');
const path = require('path');
const log4js = require('log4js');
const _ = require('underscore');

const LOG_CONFIG_FILE = path.normalize(__dirname + '/../../loggerConfig.json');
const LOG_LEVELS = ['ALL', 'TRACE', 'DEBUG', 'INFO', 'WARN', 'ERROR', 'FATAL', 'OFF'];

/**
 *  getter logger same level.
 *  default level = FATAL
 * @class Logger class.
 **/
const Logger = (function() {
	const logConfigJson = JSON.parse(fs.readFileSync(LOG_CONFIG_FILE));
	var LogLevel = 'OFF';
	/**
	 * set log level
	 * @method setLevel
	 * @param {string} level level of logs.
	 * @memberof Logger
	 */
	function setLevel(level) {
		LogLevel = getLevel(level);
		log4js.setGlobalLogLevel(level);
	}
	/**
	 * configure Logger
	 * @method configure
	 * @memberof Logger
	 */
	function configure() {
		if (LogLevel !== 'OFF') {
			log4js.configure(logConfigJson);
		}
	}
	/**
	 * addConsoleAppender  that will be logging.
	 * @method addConsoleAppender
	 * @memberof Logger
	 */
	function addConsoleAppender() {
		logConfigJson.appenders.push({
			type: 'console',
			layout: {
				type: "pattern",
				pattern: '%x{currentTime} - %-5p - %-16c - %m',
				tokens: {
					currentTime: e => new Date().toGMTString()
				}
			}
		});
		log4js.configure(logConfigJson);
	}
	/**
	 * addFileAppender  that will be logging.
	 * @method addFileAppender
	 * @memberof Logger
	 */
	function addFileAppender(filePath) {
		logConfigJson.appenders.push({
			type: 'file',
			filename: filePath,
			layout: {
				type: "pattern",
				pattern: '%x{currentTime}  - %-5p  %m',
				tokens: {
					currentTime: e => new Date().toGMTString()
				}
			}
		});
		log4js.configure(logConfigJson);
	}
	/**
	 * get logger
	 * @method getLogger
	 * @param {string} name name of logger.
	 * @return {Logger} logger as name.
	 * @memberof Logger
	 */
	function getLogger(name) {
		const logger = log4js.getLogger(name);
		logger.setLevel(LogLevel);
		return logger;
	}

	function shutdown(callback) {
		log4js.shutdown(callback);
	}

	return {
		addFileAppender: addFileAppender,
		addConsoleAppender: addConsoleAppender,
		getLogger: getLogger,
		setLevel: setLevel,
		configure: configure,
		shutdown: shutdown
	};

})();

function getLevel(level) {
	if (!_.isUndefined(level)) {
		const index = LOG_LEVELS.indexOf(level.toUpperCase());
		if (index !== -1) {
			return LOG_LEVELS[index];
		}
	}
	return 'OFF';
}



module.exports = Logger;
