// smartface/lib/interactive/sample.js
const path = require('path');
const fs = require('fs');
const _ = require('underscore');

const utility = require('../utility');
const requireJSON = require('../requireJSON');

const Sample = (function() {
	const sampleJson = requireJSON(__dirname, './sample.json');
	const Default = require('./tag');
	const SAMPLES_DIR = Default.def.SAMPLES_DIR;

	function getAvailableNames() {
		var res = [];
		_.each(sampleJson.samples, function(item) {
			res.push(item.name);
		});
		return res;
	}

	function getAvailableLink(_name) {
		const index = _.findIndex(sampleJson.samples, {
			name: _name
		});
		if (index !== -1) {
			return sampleJson.samples[index].link;
		}
	}

	function getSampleFolder(_name) { // SAMPLES_DIR subfolders.
		var res = undefined;
		if (name) {
			const sampleFolderPath = path.join(SAMPLES_DIR, _name);
			if (utility.safeControlDirectory(sampleFolderPath)) {
				res = sampleFolderPath;
			}
		}
		return res;
	}

	function getSampleNames(validator) { // need projectValidator for  realy smfc projects folder.
		return getSamplePathsOrNamesCommon(validator, true);
	}

	function getSamplePaths(validator) { // need projectValidator for real smfc projects folder.
		return getSamplePathsOrNamesCommon(validator);
	}

	function getSamplePathsOrNamesCommon(validator, isName) {
		var res = [];
		const _validator = validator || validateProjectFolder;
		if (utility.safeControlDirectory(SAMPLES_DIR)) {
			const folders = fs.readdirSync(SAMPLES_DIR);
			_.each(folders, function(item) {
				const input = path.join(SAMPLES_DIR, item);
				const whatToPush = (isName) ? item : input;
				if (_.isFunction(_validator)) {
					if (_validator(input) === true) {
						res.push(whatToPush);
					}
				} else {
					res.push(whatToPush);
				}
			});
		}
		return res;
	}

	function validateProjectFolder(input) { // validate project folder
		var answered = require('./answered');
		const convertAbsolute = utility.convertAbsolute.convert;
		var res = false;
		const projPath = convertAbsolute(input);
		if (utility.safeControlDirectory(projPath)) {
			if (answered.workSpaceType === Default.def.WORKSPACE.DESKTOP) {
				// TO DO DESKTOP ıde sfpx, data control.
			} else {
				if (utility.safeControlFile(path.join(projPath, 'config', 'project.json'), 'json')) {
					if (utility.safeControlFile(path.join(projPath, 'config', 'Android', 'PackageProfiles.xml'), 'xml')) {
						res = true;
					}
				}
			}
		}
		return res;
	}

	return {
		getAvailableNames: getAvailableNames,
		getAvailableLink: getAvailableLink,
		getSampleFolder: getSampleFolder,
		getSamplePaths: getSamplePaths,
		getSampleNames: getSampleNames
	};

})();

module.exports = Sample;
