// smartface/lib/interactive/tag.js
const platform = require('os').platform();
const _ = require('underscore');


const Default = (function() { // win32 or linux | c9 | MAC
	const sampleDir = (platform === 'win32') ? process.env.USERPROFILE +
		"/Documents/Smartface/Samples" : process.env.HOME + "/Smartface/Samples";

	const defaults = {
		WORKSPACE: {
			DESKTOP: 'IDE - Desktop',
			SMFC: 'IDE - SMF_CLOUD'
		},
		TASK: {
			CHECK: 'Check Environment & Dependencies',
			SAMPLE: 'Sample Projects',
			CREATE_IOS_PLUGIN: 'create_iOSPlugin',
			GENERATE_IOS_PLUGIN: 'generate_iOSPluginZip',
			GENERATE_ANDROID_PLUGIN: 'generate_AndroidPluginZip',
			EXIT: 'Exit'
		},
		LICENSE: {
			DEMO: 'Smartface Demo'
		},
		ANDROID: {
			JAVA: 'java',
			INPUT_APK: 'bin/SmartfacePlayer',
			INPUT_APK_x86: 'bin/SmartfacePlayer-x86.apk',
			PROFILE: "ARM - Default"
		},
		iOS: {
			INPUT_ZIP: 'bin/iOS_Player.zip'
		},
		SAMPLES_DIR: sampleDir
	};
	// returns Array of key of defaults.
	function getList(key) {
		var res = [];
		_.each(_.values(defaults[key]), function(item) {
			res.push(item);
		});
		return res;
	}

	function getPublishTasks() {
		return [defaults.TASK.ANDROID_FULL_PUBLISH, defaults.TASK.IOS_FULL_PUBLISH];
	}

	return {
		def: defaults,
		getList: getList,
		getPublishTasks: getPublishTasks
	};


})();

module.exports = Default;
