// smartface/lib/cloudBuild/wsCommunaciationCmd.js
const COMMUNICATION = {
  "CONNECT": "CONNECT",
  "COMMAND": "COMMAND",
  "FATAL": "FATAL",
  "ERROR": "ERROR",
  "INFO": "INFO",
  "DONE": "DONE",
  "FILE": "FILE"
};

const COMMAND = {
  "RUN": "RUN",
  "GET_BINARY": "GET_BINARY",
  "END": "END",
  "READY": "READY",
  "OK": "OK",
  "FILE_START": "FILE_START",
  "FILE_END": "FILE_END",
  "COMPLETED": "COMPLETED",
  "KEEPALIVE": "KEEPALIVE",
  "SAVE_BINARY": "SAVE_BINARY"
};

const bracket = "$(:)";

function parse(message) {
  var res;
  var cmds = message.split(bracket);
  if ((cmds.length === 2) && COMMUNICATION[cmds[0]]) {
    res = {};
    res.type = COMMUNICATION[cmds[0]];
    res.cmd = cmds[1];
  }
  return res;
}

function createFatal(doneMsg) {
  return COMMUNICATION.FATAL + bracket + doneMsg;
}

function createError(errMsg) {
  return COMMUNICATION.ERROR + bracket + errMsg;
}

function createInfo(infoMsg) {
  return COMMUNICATION.INFO + bracket + infoMsg;
}

function createDone(doneMsg) {
  return COMMUNICATION.DONE + bracket + doneMsg;
}

function createCmd(_cmd) {
  return COMMUNICATION.COMMAND + bracket + _cmd;
}

function createFile(size) {
  return COMMUNICATION.FILE + bracket + size;
}

module.exports = {
  COMMUNICATION: COMMUNICATION,
  COMMAND: COMMAND,
  createCmd: createCmd,
  createDone: createDone,
  createInfo: createInfo,
  createError: createError,
  createFatal: createFatal,
  createFile: createFile,
  parse: parse,
  BRACKET: bracket
};
