// smartface/lib/collector/tag.js
const SFPX = {

	ROOT_TAG: 'SMARTFACEPROJECT',
	DEVICE_ORIENTATIONS_TAG: 'DeviceOrientations',
	PROJECT_DETAILS_TAG: 'ProjectDetails',
	MAP_API_TAG: 'MapApi',
	APP_NAME_TAG: 'ApplicationName',
	IOS_KEYSTORE_TAG: 'IOSKeyStore',
	ATTR_IOS_URL_IDENTIFIER: 'IOSUrlIdentifier',
	ATRR_IOS_URL_SCHEMES: 'IOSUrlSchemes',
	ATTR_APP_NAME_TAG: 'ApplicationName',
	ATTR_SPLASH_BACKGROUND_COLOR_TAG: 'SplashScreenBackgroundColor',
	ATTR_APP_VERSION_TAG: 'Version',
	ATTR_XML_VERSION_TAG: 'XmlVersion',
	ATTR_DESCRIPTION_TAG: 'Description',
	ATTR_PORTRAIT_TAG: 'Portrait',
	ATTR_LANDSCAPE_LEFT_TAG: 'LandscapeLeft',
	ATTR_LANDSCAPE_RIGHT_TAG: 'LandscapeRight',
	ATTR_UPSIDEDOWN_TAG: 'UpsideDown',
	ATTR_ANDROID_MAP_KEY_TAG: 'AndroidMapKey'
};

const LICENSE = {
	APP_PACKAGE_LICENSE_TAG: 'ApplicationPackageLicense',
	TRIPLE_DES_TAG: 'TripleDESKeyValue',
	LICENSE_TYPE_TAG: 'LicenseType',
	LICENSE_NAME_TAG: 'LicenseName',
	PACKAGE_NAME_TAG: 'PackageName',
	PACKAGE_TYPE_TAG: 'PackageType'
};

const PACKAGE_PROFILES = {
	PROFILE_TAG: 'profile',
	ATTR_PROFILE_NAME_TAG: 'name',
	ANDROID_TAG: 'android',
	FOLDER_TAG: 'folder',
	ATTR_FOLDER_NAME_TAG: 'name',
	ATTR_PROFILE_NAME_TAG_BC: 'Name',
	ATTR_CPU_TYPE_ARM: 'ARM',
	ATTR_CPU_TYPE_X86: 'x86'
};

module.exports = {
	SFPX: SFPX,
	LICENSE: LICENSE,
	PACKAGE_PROFILES: PACKAGE_PROFILES
};
