// library/CheckLine.js
//------------------------------------------------------------------------------
//
//     This code was auto generated.
//
//     Manual changes to this file may cause unexpected behavior in your application.
//     Manual changes to this file will be overwritten if the code is regenerated.
//
//------------------------------------------------------------------------------

const extend = require('js-base/core/extend');
const FlexLayout = extend(require('sf-core/ui/flexlayout'));

function addChild(childName, ChildClass, componentInstance) {
  this.children = this.children || {};
  this.children[childName] = new ChildClass(componentInstance);
  if (this.layout)
    this.layout.addChild(this.children[childName]);
  else
    this.addChild(this.children[childName]);
}
//constructor
function CheckLine(_super, props) {
  // initalizes super class for this component scope
  _super(this, props);
  addChild.call(this, "line", $CheckLine$$Line_, this);
  addChild.call(this, "ballfl", $CheckLine$$Ballfl_, this);
}
CheckLine.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    flexProps: {
      alignItems: "CENTER"
    },
    height: 230,
    width: 70
  }
};
const CheckLine_ = FlexLayout(CheckLine);

function $CheckLine$$Line(_super, pageInstance) {
  _super(this);
}
$CheckLine$$Line.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    alpha: 0.31,
    backgroundColor: "rgba( 151, 151, 151, 1 )",
    bottom: 0,
    flexProps: {
      positionType: "ABSOLUTE"
    },
    height: null,
    marginTop: 30,
    top: 0,
    width: 1
  }
};
const $CheckLine$$Line_ = FlexLayout($CheckLine$$Line);

function $CheckLine$$Ballfl(_super, pageInstance) {
  _super(this);
  pageInstance.ballfl = this;
}
$CheckLine$$Ballfl.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    borderColor: "rgba( 208, 2, 27, 1 )",
    borderRadius: 15,
    borderWidth: 1,
    flexProps: {
      positionType: "ABSOLUTE"
    },
    height: 30,
    top: 0,
    width: 30
  }
};
const $CheckLine$$Ballfl_ = FlexLayout($CheckLine$$Ballfl);

module && (module.exports = CheckLine_);
