// library/CurrentProcedureRowfl.js
//------------------------------------------------------------------------------
//
//     This code was auto generated.
//
//     Manual changes to this file may cause unexpected behavior in your application.
//     Manual changes to this file will be overwritten if the code is regenerated.
//
//------------------------------------------------------------------------------

const extend = require('js-base/core/extend');
const FlexLayout = extend(require('sf-core/ui/flexlayout'));

const ProcedureRow = extend(require("../../components/ProcedureRow"));
const CheckLine1 = extend(require("../../components/CheckLine1"));

function addChild(childName, ChildClass, componentInstance) {
  this.children = this.children || {};
  this.children[childName] = new ChildClass(componentInstance);
  if (this.layout)
    this.layout.addChild(this.children[childName]);
  else
    this.addChild(this.children[childName]);
}
//constructor
function CurrentProcedureRowfl(_super, props) {
  // initalizes super class for this component scope
  _super(this, props);
  addChild.call(this, "procedureRow", $CurrentProcedureRowfl$$ProcedureRow_, this);
  addChild.call(this, "checkLine", $CurrentProcedureRowfl$$CheckLine_, this);
}
CurrentProcedureRowfl.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    height: 230,
    width: null,
    backgroundColor: "rgba( 255, 255, 255, 0 )"
  }
};
const CurrentProcedureRowfl_ = FlexLayout(CurrentProcedureRowfl);

function $CurrentProcedureRowfl$$ProcedureRow(_super, pageInstance) {
  _super(this);
}
$CurrentProcedureRowfl$$ProcedureRow.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    width: null,
    height: null,
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    left: 70,
    right: 0,
    top: 0,
    bottom: 0,
    flexProps: {
      flexDirection: "COLUMN_REVERSE",
      positionType: "ABSOLUTE"
    }
  }
};
const $CurrentProcedureRowfl$$ProcedureRow_ = ProcedureRow($CurrentProcedureRowfl$$ProcedureRow);

function $CurrentProcedureRowfl$$CheckLine(_super, pageInstance) {
  _super(this);
}
$CurrentProcedureRowfl$$CheckLine.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    width: 70,
    height: 230,
    left: 0,
    top: 0,
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    flexProps: {
      alignItems: "CENTER",
      positionType: "RELATIVE"
    }
  }
};
const $CurrentProcedureRowfl$$CheckLine_ = CheckLine1($CurrentProcedureRowfl$$CheckLine);

module && (module.exports = CurrentProcedureRowfl_);
