// library/EndTimePicker.js
//------------------------------------------------------------------------------
//
//     This code was auto generated.
//
//     Manual changes to this file may cause unexpected behavior in your application.
//     Manual changes to this file will be overwritten if the code is regenerated.
//
//------------------------------------------------------------------------------

const extend = require('js-base/core/extend');
const FlexLayout = extend(require('sf-core/ui/flexlayout'));
const ImageView = extend(require('sf-core/ui/imageview'));
const Label = extend(require('sf-core/ui/label'));

function addChild(childName, ChildClass, componentInstance) {
  this.children = this.children || {};
  this.children[childName] = new ChildClass(componentInstance);
  if (this.layout)
    this.layout.addChild(this.children[childName]);
  else
    this.addChild(this.children[childName]);
}
//constructor
function EndTimePicker(_super, props) {
  // initalizes super class for this component scope
  _super(this, props);
  addChild.call(this, "timeContainer", $EndTimePicker$$TimeContainer_, this);
  addChild.call(this, "line", $EndTimePicker$$Line_, this);
}
EndTimePicker.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    flexProps: {
      flexDirection: "ROW"
    },
    height: 60,
    width: 150
  }
};
const EndTimePicker_ = FlexLayout(EndTimePicker);

function $EndTimePicker$$TimeContainer(_super, pageInstance) {
  _super(this);
  addChild.call(this, "okImg", $EndTimePicker$$TimeContainer$$OkImg_, pageInstance);
  addChild.call(this, "label1", $EndTimePicker$$TimeContainer$$Label1_, pageInstance);
}
$EndTimePicker$$TimeContainer.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    flexProps: {
      flexGrow: 1
    },
    height: null,
    marginBottom: 1,
    width: null
  }
};
const $EndTimePicker$$TimeContainer_ = FlexLayout($EndTimePicker$$TimeContainer);

function $EndTimePicker$$TimeContainer$$OkImg(_super, pageInstance) {
  _super(this);
}
$EndTimePicker$$TimeContainer$$OkImg.$$styleContext = {
  classNames: ".imageView",
  userProps: {
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    flexProps: {
      positionType: "ABSOLUTE"
    },
    height: 10,
    image: "downokicon.png",
    imageFillType: "STRETCH",
    right: 0,
    top: 25,
    width: 20
  }
};
const $EndTimePicker$$TimeContainer$$OkImg_ = ImageView($EndTimePicker$$TimeContainer$$OkImg);

function $EndTimePicker$$TimeContainer$$Label1(_super, pageInstance) {
  _super(this, {
    text: "End Time"
  });
}
$EndTimePicker$$TimeContainer$$Label1.$$styleContext = {
  classNames: ".label",
  userProps: {
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    flexProps: {
      flexGrow: 1
    },
    font: {
      family: "Lato",
      size: 14
    },
    height: null,
    marginRight: 50,
    textColor: "rgba( 74, 74, 74, 1 )",
    width: null
  }
};
const $EndTimePicker$$TimeContainer$$Label1_ = Label($EndTimePicker$$TimeContainer$$Label1);

function $EndTimePicker$$Line(_super, pageInstance) {
  _super(this);
}
$EndTimePicker$$Line.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    alpha: 0.31,
    backgroundColor: "rgba( 151, 151, 151, 1 )",
    bottom: 0,
    flexProps: {
      positionType: "ABSOLUTE"
    },
    height: 1,
    left: 0,
    right: 0,
    width: null
  }
};
const $EndTimePicker$$Line_ = FlexLayout($EndTimePicker$$Line);

module && (module.exports = EndTimePicker_);
