// library/MapViewfl.js
//------------------------------------------------------------------------------
//
//     This code was auto generated.
//
//     Manual changes to this file may cause unexpected behavior in your application.
//     Manual changes to this file will be overwritten if the code is regenerated.
//
//------------------------------------------------------------------------------

const extend = require('js-base/core/extend');
const FlexLayout = extend(require('sf-core/ui/flexlayout'));
const MapView = extend(require('sf-core/ui/mapview'));
const ImageView = extend(require('sf-core/ui/imageview'));

function addChild(childName, ChildClass, componentInstance) {
  this.children = this.children || {};
  this.children[childName] = new ChildClass(componentInstance);
  if (this.layout)
    this.layout.addChild(this.children[childName]);
  else
    this.addChild(this.children[childName]);
}
//constructor
function MapViewfl(_super, props) {
  // initalizes super class for this component scope
  _super(this, props);
  addChild.call(this, "workMapView", $MapViewfl$$WorkMapView_, this);
  addChild.call(this, "redlocationImg", $MapViewfl$$RedlocationImg_, this);
  addChild.call(this, "yellowlocationImg", $MapViewfl$$YellowlocationImg_, this);
  addChild.call(this, "currentLocationImf", $MapViewfl$$CurrentLocationImf_, this);
}
MapViewfl.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    width: null,
    height: null,
    backgroundColor: "rgba( 255, 255, 255, 0 )"
  }
};
const MapViewfl_ = FlexLayout(MapViewfl);

function $MapViewfl$$WorkMapView(_super, pageInstance) {
  _super(this);
  pageInstance.workMapView = this;
}
$MapViewfl$$WorkMapView.$$styleContext = {
  classNames: ".mapView",
  userProps: {
    width: null,
    height: null,
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    flexProps: {
      flexGrow: 1
    }
  }
};
const $MapViewfl$$WorkMapView_ = MapView($MapViewfl$$WorkMapView);

function $MapViewfl$$RedlocationImg(_super, pageInstance) {
  _super(this);
  pageInstance.redlocationImg = this;
}
$MapViewfl$$RedlocationImg.$$styleContext = {
  classNames: ".imageView",
  userProps: {
    width: 30,
    height: 30,
    right: 10,
    bottom: 60,
    image: "redlocation.png",
    imageFillType: "STRETCH",
    visible: false,
    flexProps: {
      positionType: "ABSOLUTE",
      alignSelf: "AUTO"
    }
  }
};
const $MapViewfl$$RedlocationImg_ = ImageView($MapViewfl$$RedlocationImg);

function $MapViewfl$$YellowlocationImg(_super, pageInstance) {
  _super(this);
  pageInstance.yellowlocationImg = this;
}
$MapViewfl$$YellowlocationImg.$$styleContext = {
  classNames: ".imageView",
  userProps: {
    width: 30,
    height: 30,
    right: 10,
    bottom: 70,
    image: "yellowlocation.png",
    imageFillType: "STRETCH",
    visible: false,
    flexProps: {
      positionType: "ABSOLUTE",
      alignSelf: "AUTO"
    }
  }
};
const $MapViewfl$$YellowlocationImg_ = ImageView($MapViewfl$$YellowlocationImg);

function $MapViewfl$$CurrentLocationImf(_super, pageInstance) {
  _super(this);
  pageInstance.currentLocationImf = this;
}
$MapViewfl$$CurrentLocationImf.$$styleContext = {
  classNames: ".imageView",
  userProps: {
    width: 30,
    height: 30,
    right: 10,
    bottom: 50,
    image: "currentlocationicon.png",
    imageFillType: "STRETCH",
    flexProps: {
      positionType: "ABSOLUTE",
      alignSelf: "AUTO"
    }
  }
};
const $MapViewfl$$CurrentLocationImf_ = ImageView($MapViewfl$$CurrentLocationImf);

module && (module.exports = MapViewfl_);
