// library/Nofl.js
//------------------------------------------------------------------------------
//
//     This code was auto generated.
//
//     Manual changes to this file may cause unexpected behavior in your application.
//     Manual changes to this file will be overwritten if the code is regenerated.
//
//------------------------------------------------------------------------------

const extend = require('js-base/core/extend');
const FlexLayout = extend(require('sf-core/ui/flexlayout'));
const Button = extend(require('sf-core/ui/button'));
const Label = extend(require('sf-core/ui/label'));

function addChild(childName, ChildClass, componentInstance) {
  this.children = this.children || {};
  this.children[childName] = new ChildClass(componentInstance);
  if (this.layout)
    this.layout.addChild(this.children[childName]);
  else
    this.addChild(this.children[childName]);
}
//constructor
function Nofl(_super, props) {
  // initalizes super class for this component scope
  _super(this, props);
  addChild.call(this, "radioButton", $Nofl$$RadioButton_, this);
  addChild.call(this, "noLabel", $Nofl$$NoLabel_, this);
}
Nofl.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    flexProps: {
      justifyContent: "CENTER",
      flexDirection: "ROW"
    },
    height: 50,
    width: 80
  }
};
const Nofl_ = FlexLayout(Nofl);

function $Nofl$$RadioButton(_super, pageInstance) {
  _super(this);
  pageInstance.radioButton = this;
}
$Nofl$$RadioButton.$$styleContext = {
  classNames: ".button",
  userProps: {
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    borderColor: "rgba( 155, 155, 155, 1 )",
    borderRadius: 10,
    borderWidth: 1,
    flexProps: {
      alignSelf: "CENTER"
    },
    height: 20,
    width: 20
  }
};
const $Nofl$$RadioButton_ = Button($Nofl$$RadioButton);

function $Nofl$$NoLabel(_super, pageInstance) {
  _super(this, {
    text: "No"
  });
  pageInstance.noLabel = this;
}
$Nofl$$NoLabel.$$styleContext = {
  classNames: ".label",
  userProps: {
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    flexProps: {
      flexGrow: 1
    },
    font: {
      family: "Lato",
      size: 14
    },
    height: null,
    marginLeft: 10,
    textColor: "rgba( 74, 74, 74, 1 )",
    visible: true,
    width: null
  }
};
const $Nofl$$NoLabel_ = Label($Nofl$$NoLabel);

module && (module.exports = Nofl_);
