// library/Notificaitonfl.js
//------------------------------------------------------------------------------
//
//     This code was auto generated.
//
//     Manual changes to this file may cause unexpected behavior in your application.
//     Manual changes to this file will be overwritten if the code is regenerated.
//
//------------------------------------------------------------------------------

const extend = require('js-base/core/extend');
const FlexLayout = extend(require('sf-core/ui/flexlayout'));
const Label = extend(require('sf-core/ui/label'));
const Switch = extend(require('sf-core/ui/switch'));

function addChild(childName, ChildClass, componentInstance) {
  this.children = this.children || {};
  this.children[childName] = new ChildClass(componentInstance);
  if (this.layout)
    this.layout.addChild(this.children[childName]);
  else
    this.addChild(this.children[childName]);
}
//constructor
function Notificaitonfl(_super, props) {
  // initalizes super class for this component scope
  _super(this, props);
  addChild.call(this, "line", $Notificaitonfl$$Line_, this);
  addChild.call(this, "notificationContainer", $Notificaitonfl$$NotificationContainer_, this);
}
Notificaitonfl.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    height: 70,
    width: null
  }
};
const Notificaitonfl_ = FlexLayout(Notificaitonfl);

function $Notificaitonfl$$Line(_super, pageInstance) {
  _super(this);
}
$Notificaitonfl$$Line.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    alpha: 0.31,
    backgroundColor: "rgba( 61, 59, 58, 1 )",
    bottom: 0,
    flexProps: {
      positionType: "ABSOLUTE"
    },
    height: 1,
    left: 0,
    right: 0,
    width: null
  }
};
const $Notificaitonfl$$Line_ = FlexLayout($Notificaitonfl$$Line);

function $Notificaitonfl$$NotificationContainer(_super, pageInstance) {
  _super(this);
  addChild.call(this, "themeLabel", $Notificaitonfl$$NotificationContainer$$ThemeLabel_, pageInstance);
  addChild.call(this, "stylesfl", $Notificaitonfl$$NotificationContainer$$Stylesfl_, pageInstance);
  pageInstance.notificationContainer = this;
}
$Notificaitonfl$$NotificationContainer.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    bottom: 1,
    flexProps: {
      positionType: "ABSOLUTE"
    },
    height: null,
    left: 0,
    right: 0,
    top: 1,
    width: null
  }
};
const $Notificaitonfl$$NotificationContainer_ = FlexLayout($Notificaitonfl$$NotificationContainer);

function $Notificaitonfl$$NotificationContainer$$ThemeLabel(_super, pageInstance) {
  _super(this, {
    text: "Notification"
  });
  pageInstance.themeLabel = this;
}
$Notificaitonfl$$NotificationContainer$$ThemeLabel.$$styleContext = {
  classNames: ".label",
  userProps: {
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    bottom: 0,
    flexProps: {
      positionType: "ABSOLUTE"
    },
    font: {
      family: "Lato"
    },
    height: null,
    left: 1,
    right: 90,
    top: 1,
    width: null
  }
};
const $Notificaitonfl$$NotificationContainer$$ThemeLabel_ = Label($Notificaitonfl$$NotificationContainer$$ThemeLabel);

function $Notificaitonfl$$NotificationContainer$$Stylesfl(_super, pageInstance) {
  _super(this);
  addChild.call(this, "notificationSwitch", $Notificaitonfl$$NotificationContainer$$Stylesfl$$NotificationSwitch_, pageInstance);
}
$Notificaitonfl$$NotificationContainer$$Stylesfl.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    bottom: 0,
    flexProps: {
      positionType: "ABSOLUTE",
      flexDirection: "ROW",
      justifyContent: "FLEX_START",
      alignItems: "CENTER"
    },
    height: null,
    right: 0,
    top: 0,
    width: 70
  }
};
const $Notificaitonfl$$NotificationContainer$$Stylesfl_ = FlexLayout($Notificaitonfl$$NotificationContainer$$Stylesfl);

function $Notificaitonfl$$NotificationContainer$$Stylesfl$$NotificationSwitch(_super, pageInstance) {
  _super(this);
  pageInstance.notificationSwitch = this;
}
$Notificaitonfl$$NotificationContainer$$Stylesfl$$NotificationSwitch.$$styleContext = {
  classNames: ".switch",
  userProps: {
    flexProps: {
      flexGrow: 1,
      alignSelf: "AUTO"
    },
    thumbOffColor: "rgba( 208, 2, 27, 1 )",
    width: null
  }
};
const $Notificaitonfl$$NotificationContainer$$Stylesfl$$NotificationSwitch_ = Switch($Notificaitonfl$$NotificationContainer$$Stylesfl$$NotificationSwitch);

module && (module.exports = Notificaitonfl_);
