// library/Scrollfl.js
//------------------------------------------------------------------------------
//
//     This code was auto generated.
//
//     Manual changes to this file may cause unexpected behavior in your application.
//     Manual changes to this file will be overwritten if the code is regenerated.
//
//------------------------------------------------------------------------------

const extend = require('js-base/core/extend');
const FlexLayout = extend(require('sf-core/ui/flexlayout'));
const ScrollView = extend(require('sf-core/ui/scrollview'));
const ScrollViewAlign = require("sf-core/ui/scrollview/scrollview-align");

function addChild(childName, ChildClass, componentInstance) {
  this.children = this.children || {};
  this.children[childName] = new ChildClass(componentInstance);
  if (this.layout)
    this.layout.addChild(this.children[childName]);
  else
    this.addChild(this.children[childName]);
}
//constructor
function Scrollfl(_super, props) {
  // initalizes super class for this component scope
  _super(this, props);
  addChild.call(this, "uploadScroll", $Scrollfl$$UploadScroll_, this);
}
Scrollfl.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    height: 80,
    width: null
  }
};
const Scrollfl_ = FlexLayout(Scrollfl);

function $Scrollfl$$UploadScroll(_super, pageInstance) {
  _super(this, {
    align: ScrollViewAlign.HORIZONTAL
  });
  pageInstance.uploadScroll = this;
}
$Scrollfl$$UploadScroll.$$styleContext = {
  classNames: ".scrollView",
  userProps: {
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    bottom: 0,
    flexProps: {
      positionType: "ABSOLUTE"
    },
    height: null,
    ios: {
      scrollBarEnabled: true
    },
    left: 0,
    right: 0,
    top: 0,
    width: null
  }
};
const $Scrollfl$$UploadScroll_ = ScrollView($Scrollfl$$UploadScroll);

module && (module.exports = Scrollfl_);
