// library/Tickfl.js
//------------------------------------------------------------------------------
//
//     This code was auto generated.
//
//     Manual changes to this file may cause unexpected behavior in your application.
//     Manual changes to this file will be overwritten if the code is regenerated.
//
//------------------------------------------------------------------------------

const extend = require('js-base/core/extend');
const FlexLayout = extend(require('sf-core/ui/flexlayout'));
const ImageView = extend(require('sf-core/ui/imageview'));

function addChild(childName, ChildClass, componentInstance) {
  this.children = this.children || {};
  this.children[childName] = new ChildClass(componentInstance);
  if (this.layout)
    this.layout.addChild(this.children[childName]);
  else
    this.addChild(this.children[childName]);
}
//constructor
function Tickfl(_super, props) {
  // initalizes super class for this component scope
  _super(this, props);
  addChild.call(this, "tickImage", $Tickfl$$TickImage_, this);
}
Tickfl.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    width: 30,
    height: 30,
    backgroundColor: "rgba( 255, 255, 255, 0 )"
  }
};
const Tickfl_ = FlexLayout(Tickfl);

function $Tickfl$$TickImage(_super, pageInstance) {
  _super(this);
}
$Tickfl$$TickImage.$$styleContext = {
  classNames: ".imageView",
  userProps: {
    width: null,
    height: null,
    image: "tick.png",
    flexProps: {
      flexGrow: 1
    }
  }
};
const $Tickfl$$TickImage_ = ImageView($Tickfl$$TickImage);

module && (module.exports = Tickfl_);
