// library/UploadedImagefl.js
//------------------------------------------------------------------------------
//
//     This code was auto generated.
//
//     Manual changes to this file may cause unexpected behavior in your application.
//     Manual changes to this file will be overwritten if the code is regenerated.
//
//------------------------------------------------------------------------------

const extend = require('js-base/core/extend');
const FlexLayout = extend(require('sf-core/ui/flexlayout'));
const ImageView = extend(require('sf-core/ui/imageview'));

function addChild(childName, ChildClass, componentInstance) {
  this.children = this.children || {};
  this.children[childName] = new ChildClass(componentInstance);
  if (this.layout)
    this.layout.addChild(this.children[childName]);
  else
    this.addChild(this.children[childName]);
}
//constructor
function UploadedImagefl(_super, props) {
  // initalizes super class for this component scope
  _super(this, props);
  addChild.call(this, "uploadedImageview", $UploadedImagefl$$UploadedImageview_, this);
}
UploadedImagefl.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    height: 80,
    width: 80
  }
};
const UploadedImagefl_ = FlexLayout(UploadedImagefl);

function $UploadedImagefl$$UploadedImageview(_super, pageInstance) {
  _super(this);
  pageInstance.uploadedImageview = this;
}
$UploadedImagefl$$UploadedImageview.$$styleContext = {
  classNames: ".imageView",
  userProps: {
    bottom: 0,
    flexProps: {
      positionType: "ABSOLUTE"
    },
    height: null,
    imageFillType: "STRETCH",
    left: 0,
    right: 5,
    top: 0,
    width: null
  }
};
const $UploadedImagefl$$UploadedImageview_ = ImageView($UploadedImagefl$$UploadedImageview);

module && (module.exports = UploadedImagefl_);
