// library/Uploadfl.js
//------------------------------------------------------------------------------
//
//     This code was auto generated.
//
//     Manual changes to this file may cause unexpected behavior in your application.
//     Manual changes to this file will be overwritten if the code is regenerated.
//
//------------------------------------------------------------------------------

const extend = require('js-base/core/extend');
const FlexLayout = extend(require('sf-core/ui/flexlayout'));
const ImageView = extend(require('sf-core/ui/imageview'));

function addChild(childName, ChildClass, componentInstance) {
  this.children = this.children || {};
  this.children[childName] = new ChildClass(componentInstance);
  if (this.layout)
    this.layout.addChild(this.children[childName]);
  else
    this.addChild(this.children[childName]);
}
//constructor
function Uploadfl(_super, props) {
  // initalizes super class for this component scope
  _super(this, props);
  addChild.call(this, "uploadimg", $Uploadfl$$Uploadimg_, this);
}
Uploadfl.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    height: 100,
    width: 100
  }
};
const Uploadfl_ = FlexLayout(Uploadfl);

function $Uploadfl$$Uploadimg(_super, pageInstance) {
  _super(this);
  pageInstance.uploadimg = this;
}
$Uploadfl$$Uploadimg.$$styleContext = {
  classNames: ".imageView",
  userProps: {
    flexProps: {
      flexGrow: 1
    },
    height: null,
    image: "uploadicon.png",
    width: null
  }
};
const $Uploadfl$$Uploadimg_ = ImageView($Uploadfl$$Uploadimg);

module && (module.exports = Uploadfl_);
