// library/UploadDrawfl.js
//------------------------------------------------------------------------------
//
//     This code was auto generated.
//
//     Manual changes to this file may cause unexpected behavior in your application.
//     Manual changes to this file will be overwritten if the code is regenerated.
//
//------------------------------------------------------------------------------

const extend = require('js-base/core/extend');
const FlexLayout = extend(require('sf-core/ui/flexlayout'));

function addChild(childName, ChildClass, componentInstance) {
  this.children = this.children || {};
  this.children[childName] = new ChildClass(componentInstance);
  if (this.layout)
    this.layout.addChild(this.children[childName]);
  else
    this.addChild(this.children[childName]);
}
//constructor
function UploadDrawfl(_super, props) {
  // initalizes super class for this component scope
  _super(this, props);
  addChild.call(this, "verLine", $UploadDrawfl$$VerLine_, this);
  addChild.call(this, "horLine", $UploadDrawfl$$HorLine_, this);
}
UploadDrawfl.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    alpha: 0.31,
    backgroundColor: "rgba( 255, 255, 255, 0 )",
    borderColor: "rgba( 151, 151, 151, 0.5 )",
    borderWidth: 1,
    flexProps: {
      alignSelf: "AUTO",
      alignItems: "CENTER",
      justifyContent: "CENTER"
    },
    height: 80,
    width: 80
  }
};
const UploadDrawfl_ = FlexLayout(UploadDrawfl);

function $UploadDrawfl$$VerLine(_super, pageInstance) {
  _super(this);
}
$UploadDrawfl$$VerLine.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    alpha: 0.5,
    backgroundColor: "rgba( 151, 151, 151, 1 )",
    flexProps: {
      positionType: "ABSOLUTE"
    },
    height: 40,
    width: 2
  }
};
const $UploadDrawfl$$VerLine_ = FlexLayout($UploadDrawfl$$VerLine);

function $UploadDrawfl$$HorLine(_super, pageInstance) {
  _super(this);
}
$UploadDrawfl$$HorLine.$$styleContext = {
  classNames: ".flexLayout",
  userProps: {
    alpha: 0.5,
    backgroundColor: "rgba( 151, 151, 151, 1 )",
    flexProps: {
      positionType: "ABSOLUTE"
    },
    height: 1,
    width: 40
  }
};
const $UploadDrawfl$$HorLine_ = FlexLayout($UploadDrawfl$$HorLine);

module && (module.exports = UploadDrawfl_);
