// smartface/lib/checkpublishrules/index.js
const utility = require('../utility');
const error = require('../error/index');

const PUBLISH_TYPES = {
	LOCAL: 'local',
	TESTING_DIST: 'test',
	RAU: 'rau',
	ENTERPRISE_DIST: 'enterprise',
	STORE_SUBMISSION: "storeSubmission"
};
const ALLOWED_PUBLISH_TYPES = [
	PUBLISH_TYPES.LOCAL,
	PUBLISH_TYPES.TESTING_DIST,
	PUBLISH_TYPES.RAU,
	PUBLISH_TYPES.ENTERPRISE_DIST,
	PUBLISH_TYPES.STORE_SUBMISSION
];

function run(args) {
	if (utility.isValidTask(args.task, "checkProject")) {
		checkArgs(args);
		error.getErrorReport(args, function(err, report) {
			if (err) return console.error(err);
			console.log(report);
		});
	}

	/**
	 * If you need to check for another argument, don't forget to modify:
	 * - lib/task.json
	 * - lib/error/error.json
	 */
	function checkArgs(args) {
		var publishType = args.publishType;
		if (publishType && ALLOWED_PUBLISH_TYPES.indexOf(publishType) == -1) {
			utility.killProcess(error.getObject("checkProject.publishType", {
				msg: {
					type: args.publishType,
				}
			}));
		}
	}
}

exports.run = run;
